applications/FluidDynamicsApplication/custom_elements/fluid_adjoint_element.cpp

```diff
@@ -0,0 +1,717 @@
+
+    using Derivatives = typename TAdjointElementData::SensitivityDerivatives;
+
+    typename Derivatives::Data  element_data(*this, *mpConstitutiveLaw, rCurrentProcessInfo);
+    typename Derivatives::Shape derivative(element_data);
+
+    VectorF residual;
+
+    for (IndexType g = 0; g < Ws.size(); ++g) {
+        const Vector& N = row(Ns, g);
+        const Matrix& dNdX = dNdXs[g];
+        const double W = Ws[g];
+
+        element_data.CalculateGaussPointData(W, N, dNdX);
+
+        Geometry<Point>::JacobiansType J;
+        this->GetGeometry().Jacobian(J, integration_method);
+        const auto& DN_De = this->GetGeometry().ShapeFunctionsLocalGradients(integration_method);
+
+        GeometricalSensitivityUtility::ShapeFunctionsGradientType dNdX_derivative;
+        const Matrix& rJ = J[g];
+        const Matrix& rDN_De = DN_De[g];
+        const double inv_detJ = 1.0 / MathUtils<double>::DetMat(rJ);
+        GeometricalSensitivityUtility geom_sensitivity(rJ, rDN_De);
+
+        ShapeParameter deriv;
+        IndexType row = 0;
+        for (deriv.NodeIndex = 0; deriv.NodeIndex < TNumNodes; ++deriv.NodeIndex) {
+            for (deriv.Direction = 0; deriv.Direction < TDim; ++deriv.Direction) {
+                double detJ_derivative;
+                geom_sensitivity.CalculateSensitivity(deriv, detJ_derivative, dNdX_derivative);
+                const double W_derivative = detJ_derivative * inv_detJ * W;
+
+                derivative.CalculateGaussPointResidualsDerivativeContributions(residual, deriv.NodeIndex, deriv.Direction, W, N, dNdX, W_derivative, detJ_derivative, dNdX_derivative);
+                AssembleSubVectorToMatrix(rOutput, row++, residual);
+            }
+        }
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::CalculateGeometryData(
+    Vector& rGaussWeights,
+    Matrix& rNContainer,
+    ShapeFunctionDerivativesArrayType& rDN_DX,
+    const GeometryData::IntegrationMethod& rIntegrationMethod) const
+{
+    const auto& r_geometry = this->GetGeometry();
+    const IndexType number_of_gauss_points =
+        r_geometry.IntegrationPointsNumber(rIntegrationMethod);
+
+    Vector DetJ;
+    r_geometry.ShapeFunctionsIntegrationPointsGradients(rDN_DX, DetJ, rIntegrationMethod);
+
+    if (rNContainer.size1() != number_of_gauss_points || rNContainer.size2() != TNumNodes) {
+        rNContainer.resize(number_of_gauss_points, TNumNodes, false);
+    }
+    rNContainer = r_geometry.ShapeFunctionsValues(rIntegrationMethod);
+
+    const auto& IntegrationPoints = r_geometry.IntegrationPoints(rIntegrationMethod);
+
+    if (rGaussWeights.size() != number_of_gauss_points) {
+        rGaussWeights.resize(number_of_gauss_points, false);
+    }
+
+    for (IndexType g = 0; g < number_of_gauss_points; ++g) {
+        rGaussWeights[g] = DetJ[g] * IntegrationPoints[g].Weight();
+    }
+}
```

Comment:
Already in base class.

Reply:
Because `TAdjointElementData` is different from `TElementData`. I was not able to use same `TElementData` from fluid elements because structure is different.